# DSCI 320: Project Milestone 1 
## Global YouTube Statistics 2023

This data visualization project will analyse key trends and patterns derived from the Global YouTube Statistics from 2023 dataset, available on Kaggle:
https://www.kaggle.com/datasets/nelgiriyewithana/global-youtube-statistics-2023/data

# PART I: Initial Exploration

### Data Abstraction
The dataset we chose for this project includes 25+ attributes. From this, we chose a subset of 13 relevant attributes for our initial exploration. These are listed in the table below: 

| data attribute                 | semantics                                                  | data attribute type | cardinality        |
|--------------------------------|------------------------------------------------------------|---------------------|-------------------|
| subscribers                    | number of subscribers to channel                          | quantitative        | Min: 12300000, Max: 245000000 |
| video views                    | total views across all videos on channel                  | quantitative        | Min: 0, Max: 2.28e+11 |
| category                       | category/niche of channel                                 | nominal             | 18                |
| uploads                        | total number of videos uploaded on channel                | quantitative        | Min: 0, Max: 301308 |
| country                        | country where YouTube channel originates                   | nominal             | 48                |
| channel type                   | type of YouTube channel (e.g., individual, brand)          | nominal             | 14                |
| lowest yearly earnings         | lowest estimated yearly earnings from channel             | quantitative        | Min: 0, Max: 6900000 |
| highest yearly earnings        | highest estimated yearly earnings from channel            | quantitative        | Min: 0, Max: 110600000 |
| created year                   | year when YouTube channel was created                     | temporal            | Min: 1970, Max: 2022 |
| created month                  | month YouTube channel was created                         | ordinal/temporal    | 12                |
| gross tertiary education (%)   | percentage of population enrolled in tertiary education   | quantitative        | Min: 7.6, Max: 113.1 (check max value) |
| population                     | total population of country                               | quantitative        | Min: 2.025060e+05, Max: 1.397715e+09 |
| unemployment rate              | unemployment rate in the country                           | quantitative        | Min: 0.75, Max: 14.72 |



### Introduction
From this project, we hope to learn about some of the factors that may or may not be correlated with each other in the Youtube space. With the large amount of creators and video uploads, this dataset can give interesting insights into Youtube video trends, metrics, user behaviour, and more. This information can be beneficial for creators, businesses, and users. The main intended audience would be creators, advertising companies and brand makers. Understanding popular content and niches will uncover opportunities for effective advertising and thus, monetization. 


## Exploratory Data Analysis

In [1]:
import altair as alt
import pandas as pd
import os

In [2]:
data = pd.read_csv("Data/Global YouTube Statistics.csv", encoding='latin-1')
data.head()

rank                    Youtuber  subscribers   video views  \
0     1                    T-Series    245000000  2.280000e+11   
1     2              YouTube Movies    170000000  0.000000e+00   
2     3                     MrBeast    166000000  2.836884e+10   
3     4  Cocomelon - Nursery Rhymes    162000000  1.640000e+11   
4     5                   SET India    159000000  1.480000e+11   

           category                       Title  uploads        Country  \
0             Music                    T-Series    20082          India   
1  Film & Animation               youtubemovies        1  United States   
2     Entertainment                     MrBeast      741  United States   
3         Education  Cocomelon - Nursery Rhymes      966  United States   
4             Shows                   SET India   116536          India   

  Abbreviation   channel_type  ...  subscribers_for_last_30_days  \
0           IN          Music  ...                     2000000.0   
1           US          Games  ...                           NaN   
2           US  Entertainment  ...                     8000000.0   
3           US      Education  ...                     1000000.0   
4           IN  Entertainment  ...                     1000000.0   

   created_year  created_month  created_date  \
0        2006.0            Mar          13.0   
1        2006.0            Mar           5.0   
2        2012.0            Feb          20.0   
3        2006.0            Sep           1.0   
4        2006.0            Sep          20.0   

   Gross tertiary education enrollment (%)    Population  Unemployment rate  \
0                                     28.1  1.366418e+09               5.36   
1                                     88.2  3.282395e+08              14.70   
2                                     88.2  3.282395e+08              14.70   
3                                     88.2  3.282395e+08              14.70   
4                                     28.1  1.366418e+09               5.36   

   Urban_population   Latitude  Longitude  
0       471031528.0  20.593684  78.962880  
1       270663028.0  37.090240 -95.712891  
2       270663028.0  37.090240 -95.712891  
3       270663028.0  37.090240 -95.712891  
4       471031528.0  20.593684  78.962880  

[5 rows x 28 columns]

In [3]:
print(data.shape)
print(data.info())

(995, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 995 entries, 0 to 994
Data columns (total 28 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   rank                                     995 non-null    int64  
 1   Youtuber                                 995 non-null    object 
 2   subscribers                              995 non-null    int64  
 3   video views                              995 non-null    float64
 4   category                                 949 non-null    object 
 5   Title                                    995 non-null    object 
 6   uploads                                  995 non-null    int64  
 7   Country                                  873 non-null    object 
 8   Abbreviation                             873 non-null    object 
 9   channel_type                             965 non-null    object 
 10  video_views_rank                        

In [4]:
data.isnull().sum()

rank                                         0
Youtuber                                     0
subscribers                                  0
video views                                  0
category                                    46
Title                                        0
uploads                                      0
Country                                    122
Abbreviation                               122
channel_type                                30
video_views_rank                             1
country_rank                               116
channel_type_rank                           33
video_views_for_the_last_30_days            56
lowest_monthly_earnings                      0
highest_monthly_earnings                     0
lowest_yearly_earnings                       0
highest_yearly_earnings                      0
subscribers_for_last_30_days               337
created_year                                 5
created_month                                5
created_date 

In [5]:
selected_columns = ['subscribers', 'video views', 'category', 'uploads',
                    'Country', 'channel_type','lowest_yearly_earnings',
                    'highest_yearly_earnings','created_year','created_month',
                   'Gross tertiary education enrollment (%)','Population','Unemployment rate' ]
filtered_data = data[selected_columns]
filtered_data.head()

subscribers   video views          category  uploads        Country  \
0    245000000  2.280000e+11             Music    20082          India   
1    170000000  0.000000e+00  Film & Animation        1  United States   
2    166000000  2.836884e+10     Entertainment      741  United States   
3    162000000  1.640000e+11         Education      966  United States   
4    159000000  1.480000e+11             Shows   116536          India   

    channel_type  lowest_yearly_earnings  highest_yearly_earnings  \
0          Music              6800000.00             1.084000e+08   
1          Games                    0.04             5.800000e-01   
2  Entertainment              4000000.00             6.470000e+07   
3      Education              5900000.00             9.480000e+07   
4  Entertainment              5500000.00             8.750000e+07   

   created_year created_month  Gross tertiary education enrollment (%)  \
0        2006.0           Mar                                     28.1   
1        2006.0           Mar                                     88.2   
2        2012.0           Feb                                     88.2   
3        2006.0           Sep                                     88.2   
4        2006.0           Sep                                     28.1   

     Population  Unemployment rate  
0  1.366418e+09               5.36  
1  3.282395e+08              14.70  
2  3.282395e+08              14.70  
3  3.282395e+08              14.70  
4  1.366418e+09               5.36

### Data Wrangling
We have chosen to remove all null-values as the size of the data-set stays sufficient even after doing so. In the original dataset there were 28 attributes and 995 observations. After filtering for the desired attributes and removing all null-values we have 13 attributes and 824 observations

In [6]:
clean_data = filtered_data.dropna()
print(clean_data.shape)
print(clean_data.info())

(824, 13)
<class 'pandas.core.frame.DataFrame'>
Index: 824 entries, 0 to 994
Data columns (total 13 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   subscribers                              824 non-null    int64  
 1   video views                              824 non-null    float64
 2   category                                 824 non-null    object 
 3   uploads                                  824 non-null    int64  
 4   Country                                  824 non-null    object 
 5   channel_type                             824 non-null    object 
 6   lowest_yearly_earnings                   824 non-null    float64
 7   highest_yearly_earnings                  824 non-null    float64
 8   created_year                             824 non-null    float64
 9   created_month                            824 non-null    object 
 10  Gross tertiary education enrollment (%)  824 

In [7]:
#to view in voyager
#clean_data.to_csv("/Users/KAROLINE/Desktop/DSCI 320/Project/cleaned_data_YouTube.csv", index=False)

### Cardinality

In [8]:
clean_data.agg(
    {
        "subscribers":['min', 'max'],
        "video views":['min', 'max'],
        "category": ['nunique'],
        "uploads": ['min', 'max'],
        "Country": ['nunique'],
        "channel_type": ['nunique'],
        "lowest_yearly_earnings":['min', 'max'],
        "highest_yearly_earnings":['min', 'max'],
        "created_year":['nunique'],
        "created_year":['min', 'max'],
        "created_month":['nunique'],
        "Gross tertiary education enrollment (%)":['min', 'max'],
        "Population":['min', 'max'],
        "Unemployment rate":['min', 'max'],

    }
)

subscribers   video views  category   uploads  Country  channel_type  \
min       12300000.0  0.000000e+00       NaN       0.0      NaN           NaN   
max      245000000.0  2.280000e+11       NaN  301308.0      NaN           NaN   
nunique          NaN           NaN      18.0       NaN     48.0          14.0   

         lowest_yearly_earnings  highest_yearly_earnings  created_year  \
min                         0.0                      0.0        1970.0   
max                   6900000.0              110600000.0        2022.0   
nunique                     NaN                      NaN           NaN   

         created_month  Gross tertiary education enrollment (%)    Population  \
min                NaN                                      7.6  2.025060e+05   
max                NaN                                    113.1  1.397715e+09   
nunique           12.0                                      NaN           NaN   

         Unemployment rate  
min                   0.75  
max                  14.72  
nunique                NaN

## Univariate summaries

### Numeric summaries

In [9]:
#pd.set_option('display.float_format', lambda x: '%.0f' % x) #remove # if you want to see values as integers
#pd.reset_option('display.float_format')
clean_data.describe().round(1)

subscribers   video views   uploads  lowest_yearly_earnings  \
count        824.0  8.240000e+02     824.0                   824.0   
mean    23069296.1  1.149078e+10   10964.2                456751.2   
std     18172673.0  1.509424e+10   37279.4                794091.9   
min     12300000.0  0.000000e+00       0.0                     0.0   
25%     14500000.0  4.351711e+09     316.2                 64050.0   
50%     17600000.0  7.881591e+09     966.5                192000.0   
75%     24225000.0  1.385086e+10    3597.2                489800.0   
max    245000000.0  2.280000e+11  301308.0               6900000.0   

       highest_yearly_earnings  created_year  \
count                    824.0         824.0   
mean                 7313899.7        2012.2   
std                 12723050.5           4.3   
min                        0.0        1970.0   
25%                  1000000.0        2009.0   
50%                  3100000.0        2013.0   
75%                  7825000.0        2015.0   
max                110600000.0        2022.0   

       Gross tertiary education enrollment (%)    Population  \
count                                    824.0  8.240000e+02   
mean                                      63.4  4.327673e+08   
std                                       25.9  4.755086e+08   
min                                        7.6  2.025060e+05   
25%                                       36.3  8.313280e+07   
50%                                       67.0  3.282395e+08   
75%                                       88.2  3.282395e+08   
max                                      113.1  1.397715e+09   

       Unemployment rate  
count              824.0  
mean                 9.2  
std                  4.9  
min                  0.8  
25%                  5.4  
50%                  8.9  
75%                 14.7  
max                 14.7

### Categorical valriables

In [10]:
for column in clean_data.select_dtypes(include=['object']):
    print(clean_data[column].value_counts())

category
Entertainment            211
Music                    180
People & Blogs           102
Gaming                    80
Comedy                    66
Education                 42
Film & Animation          36
Howto & Style             32
News & Politics           26
Science & Technology      14
Shows                     12
Sports                    11
Pets & Animals             3
Trailers                   2
Nonprofits & Activism      2
Movies                     2
Autos & Vehicles           2
Travel & Events            1
Name: count, dtype: int64
Country
United States           290
India                   161
Brazil                   61
United Kingdom           42
Mexico                   33
Indonesia                27
Spain                    21
Thailand                 18
Russia                   16
South Korea              15
Canada                   15
Argentina                13
Colombia                 11
Philippines              10
Saudi Arabia              9
Ukraine        

In [11]:
nominal_columns = clean_data.select_dtypes(include=['object']).columns
charts = []
for column in nominal_columns:
    category_counts = clean_data[column].value_counts().reset_index()
    category_counts.columns = [column, 'count']
    chart = alt.Chart(category_counts).mark_bar().encode(
        y=alt.Y(column, title=column).sort('-x'),
        x=alt.X('count'),
        tooltip=[column, 'count']
    ).properties(
        title=f'Count of Categories in the {column} Attribute'
    )
    charts.append(chart)
#charts[0]
alt.vconcat(*charts)

alt.VConcatChart(...)

## Quantitative variables

In [12]:
quantitative_columns = clean_data.select_dtypes(include=['float64','int64']).columns
quantitative_columns

Index(['subscribers', 'video views', 'uploads', 'lowest_yearly_earnings',
       'highest_yearly_earnings', 'created_year',
       'Gross tertiary education enrollment (%)', 'Population',
       'Unemployment rate'],
      dtype='object')

## Multivariate summaries
### Correlation 

In [13]:
cor_data = (data.drop(columns=['rank','video_views_rank','country_rank','channel_type_rank',
                               'video_views_for_the_last_30_days', 'subscribers_for_last_30_days', 
                               'Urban_population', 'Latitude', 'Longitude','Unemployment rate',
                               'lowest_monthly_earnings','highest_monthly_earnings' ])
            .corr().stack()
            .reset_index()     # The stacking results in an index on the correlation values, we need the index as normal columns for Altair
            .rename(columns={0: 'correlation', 'level_0': 'variable1', 'level_1': 'variable2'}))
cor_data['correlation_label'] = cor_data['correlation'].map('{:.2f}'.format)  # Round to 2 decimal
cor_data.head()

ValueError: could not convert string to float: 'T-Series'

In [14]:
alt.Chart(cor_data).mark_rect().encode(
    alt.X('variable1:O'),
    alt.Y('variable2:O'),
    alt.Color('correlation'),
    alt.Tooltip(['correlation_label']),
).interactive().properties(width=300, height=300)

NameError: name 'cor_data' is not defined

It is obvious that the highest and lowest yearly earnings correlate with each other. Lets investigate Video views and subscribers, as well as the earnings:

In [15]:
alt.Chart(clean_data).mark_circle().encode(
    alt.X(alt.repeat("column"), type='quantitative'),
    alt.Y(alt.repeat("row"), type='quantitative'),
).properties(
    width=100,
    height=100
).repeat(
    row= ['subscribers', 'video views', 'highest_yearly_earnings'],
    column=['subscribers', 'video views', 'highest_yearly_earnings'],
)

alt.RepeatChart(...)

In [16]:
alt.Chart(clean_data).mark_bar().encode(
    alt.Y('subscribers:Q'),
    alt.X('created_year:O'),
    color = 'channel_type'
)

alt.Chart(...)

# PART II: Project Scope 

## Task Analysis 

- Distribution → What is the distribution of subscribers for each YouTube channel type?  
- Filter → What are the top Earners Among YouTube Channels by category on a Yearly Basis? 
- Find Extremum → Which country generates the most video views yearly?
- Correlate → Is there a trend between video upload months and number of views over the years? 
- Cluster → Is there a cluster of average income a youtube channel type makes by country?
- Compute derived value → What is the average unemployment rate in the countries with the highest amount of uploads.




# Part III: Visualization Ideas 

In [17]:
from IPython.display import Image 
img1 = Image(url="img1.png", width = 500, height = 700)
img1

"What is the distribution of subscribers for each Youtube Channel type?"

A simple bar graph provides all the necessary information without distractors. It is relatively easy to compare the counts of subscribers for each category. The second visualization provides a more artistic view, with each color representing a distinct category. The size of the colored portion provides details of the subscribers count. This visualization is great for simple proportion comparison between categories. However, it might be hard to compare categories far away from each other. The third is a simple line chart. It appears more precise and cleaner than a bar graph. 

We would select a bar graph to answer this question, as it quickly highlights the channel types that have higher or lower distribution of subscribers.

In [31]:
img1final = alt.Chart(clean_data).mark_bar().encode(
    alt.X('subscribers:Q', title = 'Subscribers'),
    alt.Y('channel_type:N', title = 'Channel Type').sort('-x')
).properties(title = 'Subscriber Counts by Channel Type')
img1final

alt.Chart(...)

In [18]:
img2 = Image(url="img2.png", width = 500, height = 700)
img2

"Filter: What are the top Earners Among YouTube Channels by category on a Yearly Basis?"

The first visualization is a stacked bar graph. It is great for comparing the net income of youtube channels over the years.
The second is a pie chart for every year. You will notice that the area for 2016 is larger than 2015, since Youtube channels made more money in 2016 than 2015.
Lastly, we have the normalized bar graph. This is great for comparing the proportion of income for every channel by year. This visualization provides the most amount of information in a simple and understandable manner.

In [34]:
# normalized bar chart

In [35]:
img3 = Image(url="img3.png", width = 500, height = 700)
img3

"Correlate: Is there a trend between video upload months and number of views over the years?"



# PART IV: Next Steps

1. Become more familiar with the data by reviewing our milestone 1 tasks as a group
2.
3.
4.
5.